

# Recipe #3:  Synonym Candidates



In [24]:
import sys
!conda install --yes --prefix {sys.prefix} elasticsearch jsonlines jsonpath-ng python-slugify

import pandas
from extract import extract, parse_json
from synonym.transform import reshape, preload
from index import create_index, populate_index, query_index, index_mapping, query_template, peek


Solving environment: done

# All requested packages already installed.




## Source data 


In [25]:
source = extract('./data/real_estate_queries.tsv')

# peek
source.head(5)

user                                              query  \
0  1337           national real estate settlement services   
1  1337           national real estate settlement services   
2  1337  sunbury pennsylvania real estate settlement se...   
3  1337                             integrated real estate   
4  1337                             integrated real estate   

             timestamp  position                       document         state  
0  2006-03-14 15:59:13       1.0         http://www.realtms.com           NaN  
1  2006-03-14 15:59:13       7.0                http://dmoz.org           NaN  
2  2006-03-14 16:06:28      14.0       http://pa.optimuslaw.com  pennsylvania  
3  2006-03-27 14:52:29       1.0  http://www.integratedreal.com           NaN  
4  2006-03-27 14:52:29       2.0         http://www.irisnet.net           NaN


##  Transform to fit the recipe


In [26]:
reshaped = reshape(source)

# peek
reshaped.head(50)

document  \
0   http://0-realestate.yahoo.com.portia.nesl.edu   
1                      http://0132.deltagroup.com   
2                    http://1046391.wkg0iqbv.info   
3                     http://1133.palletmarket.cz   
4                        http://123relocation.com   
5                   http://126.realnetsystems.com   
6                         http://1603.n0jp7p.info   
7                           http://170.142.31.248   
8                            http://206.173.89.38   
9                            http://206.173.89.41   
10                          http://209.15.138.224   
11                          http://209.215.175.77   
12                           http://209.51.193.83   
13                        http://21.bataccess.com   
14                          http://211.kv446.info   
15     http://26701.real-estate.fugos.kalmykia.ru   
16                        http://2902.44w060.info   
17                    http://2993654.wlh0koy.info   
18                    http://326839.puc2e7lu.info   
19                        http://3492.74s3zv.info   
20                  http://360.home3.bls44xx.info   
21                              http://360dof.com   
22             http://36501.real-estate.nejdc.org   
23                       http://4156.yubp8jv.info   
24             http://4158.aquarium-tuinvijver.be   
25                        http://4385.k907i2.info   
26                    http://450.yourkwoffice.com   
27                  http://4867.misk2.jdjd33.info   
28                            http://4brevard.com   
29                        http://5116.qa1abr.info   
30                    http://525228.tl2b66dt.info   
31             http://5328.aquarium-tuinvijver.be   
32                         http://564.b5y31k.info   
33                   http://5770.10.ggjhdv32.info   
34                   http://5832.10.ggjhdv32.info   
35                    http://630202.b7ljvc6r.info   
36                            http://64.41.125.45   
37                        http://7520.4dbwxb.info   
38                   http://9259.10.ggjhdv32.info   
39                   http://a257.g.akamaitech.net   
40                    http://ab.virtual-agent.com   
41               http://abc7news.doityourself.com   
42                         http://abclocal.go.com   
43                          http://abcnews.go.com   
44                       http://about-the-web.com   
45                    http://absolutemichigan.com   
46                   http://academics.memphis.edu   
47          http://acronyms.thefreedictionary.com   
48                    http://actionskydancers.com   
49               http://adamscountyrealestate.net   

                                                query  
0                                 [yahoo real estate]  
1              [real estate for sale johnsonville tn]  
2                           [pureto rico real estate]  
3                            [bakersfied real estate]  
4                            [las cruces real estate]  
5   [real estate grand junction co, grand junction...  
6                         [practice real estate test]  
7   [real estate assessment data, real estate asse...  
8                              [tuson az real estate]  
9                      [panama real estate carribean]  
10  [el salvador real estate, real estate monterre...  
11                             [urbinati real estate]  
12  [franklin county ohio real estate, franklin co...  
13                   [karen real estate in milwaukee]  
14                  [branford conneticut real estate]  
15          [green mountain real estate mill city or]  
16                          [real estate calculatere]  
17                             [lflorida real estate]  
18                [real estate jacxksonville florida]  
19                          [real estate cobleslkill]  
20                          [real estate in bromx ny]  
21                               [real estate flyers]  
22                [real estate in buckingham flor

E.g. 126.realnetsystems.com  received search clicks for multiple "grand junction" queries.

In [27]:
load_ready = preload(reshaped)

# peek
[print(j) for j in load_ready[0:10]]


{'index': {'_index': 'synonym', '_id': 'http://0-realestate.yahoo.com.portia.nesl.edu'}}
{'document': 'http://0-realestate.yahoo.com.portia.nesl.edu', 'query': ['yahoo real estate']}
{'index': {'_index': 'synonym', '_id': 'http://0132.deltagroup.com'}}
{'document': 'http://0132.deltagroup.com', 'query': ['real estate for sale johnsonville tn']}
{'index': {'_index': 'synonym', '_id': 'http://1046391.wkg0iqbv.info'}}
{'document': 'http://1046391.wkg0iqbv.info', 'query': ['pureto rico real estate']}
{'index': {'_index': 'synonym', '_id': 'http://1133.palletmarket.cz'}}
{'document': 'http://1133.palletmarket.cz', 'query': ['bakersfied real estate']}
{'index': {'_index': 'synonym', '_id': 'http://123relocation.com'}}
{'document': 'http://123relocation.com', 'query': ['las cruces real estate']}


[None, None, None, None, None, None, None, None, None, None]


## Load the index


In [29]:
create_index(index='synonym', mapping_file='synonym/index_definition.json')

index_mapping(index='synonym')

Mapping for  synonym
{
  "synonym": {
    "mappings": {
      "properties": {
        "document": {
          "type": "keyword"
        },
        "query": {
          "type": "search_as_you_type",
          "analyzer": "query_log_analyzer",
          "max_shingle_size": 3
        },
        "state": {
          "type": "keyword"
        }
      }
    }
  }
}


In [30]:
response = populate_index(body=load_ready)


## Query for related searches


In [31]:
peek(filename='synonym/query_template.mustache')

{
    "size": 0,
    "query": {
      "multi_match": {
        "query": "{{text}}",
        "type": "phrase",
        "fields": [
           "query",
           "query._2gram",
           "query._3gram"
         ],
         "minimum_should_match": "100%"
       }
    },
    "aggs": {
      "significant_queries_1": {
        "significant_text": {
          "field": "query",
          "size": "{{^size}}10{{/size}}{{size}}"
        }
      },
      "significant_queries_2": {
        "significant_text": {
          "field": "query._2gram",
          "size": "{{^size}}10{{/size}}{{size}}"
        }
      },
      "significant_queries_3": {
        "significant_text": {
          "field": "query._3gram",
          "size": "{{^size}}10{{/size}}{{size}}"
        }
      }
    }
}



In [35]:
result = query_template(index='synonym', 
               template_file='synonym/query_template.mustache',
               params={ 'text': 'license', 'size': 15 }
              )

synonym_candidates = parse_json(result, jsonpath="aggregations.*.buckets[*]")
df = pandas.DataFrame(synonym_candidates).sort_values(by='score', ascending=False)

df.head(25)


key  doc_count         score  bg_count
15                    license        121  12255.000000         0
16                conditional         11    101.198347         0
17               requirements         19    100.574164         3
0         real_estate license        100     83.710129       100
18                   licenses          8     53.508367         1
19                    renewal          9     33.828222         2
20                 department         18     24.507678        11
21                  licensing         21     22.899051        16
1                  license in         25     20.927532        25
22                    licence         10     20.844888         4
23                examination          7     20.451130         2
24                    classes         36     20.171975        53
30     real_estate license in         22     18.416228        22
25                      class         12     17.121196         7
26                       exam         21     15.208183        24
27                 commission         28     15.031093        43
28                  florida's          6     15.018236         2
29                  assistant          6     15.018236         2
31      a real_estate license         14     11.719418        14
2       real_estate licensing         21     11.476643        32
3         real_estate classes         32     11.125728        76
4        license requirements         13     10.882317        13
5               department of         18     10.807180        25
6         real_estate courses         23     10.209882        43
32  department of real_estate         15      9.886943        19